automated/android/kunit/kunit.sh

```diff
@@ -0,0 +1,134 @@
+    ../../utils/upload-to-squad.sh -a "${ATTACHMENT}" -u "${SQUAD_UPLOAD_URL}"
+}
+
+function parse_kunit_log(){
+    local f_kunit_log="${1}"
+    local f_kunit_stub_log="${DIR_TEST_LOGS}/kunit_stub.log"
+
+    if [ -z "${f_kunit_log}" ] || [ ! -f "${f_kunit_log}" ]; then
+        echo "KUnit log does not exist"
+        return
+    fi
+    # grep the stub log to a single file and parsing the results
+    # 20:43:20 stub: soc-utils-test.soc-utils#test_snd_soc_params_to_bclk: PASSED (0ms)
+    # 00:21:09 stub: kunit-example-test.example_init#example_init_test: PASSED (0ms)
+    #  | cut -d: -f4- \            # kunit-example-test.example_init#example_init_test: PASSED (0ms)
+    #  | tr -d ':' \               # kunit-example-test.example_init#example_init_test PASSED (0ms)
+    #  | awk '{print $1, $2}' \    # kunit-example-test.example_init#example_init_test PASSED
+    #  | sort | uniq \             # to filter out the duplication of FAILURE in Result Summary part
+    grep "stub:" "${f_kunit_log}" \
+        | cut -d: -f4- \
+        | tr -d ':' \
+        | awk '{print $1, $2}' \
+        | sort | uniq \
+        > "${f_kunit_stub_log}"
+    while read -r line; do
+        # kunit-example-test.example_init#example_init_test PASSED
+        # kunit-example-test.example#example_skip_test IGNORED
+        # soc-utils-test#soc-utils-test FAILURE
+        test_case_name=$(echo "${line}"|awk '{print $1}')
+        test_case_result=$(echo "${line}"|awk '{print $2}')
+
+        # reformat the test case name to avoid potential confusions
+        # being caused by some special characters
+        test_case_name=$(echo "${test_case_name}" \
+                            | tr -c '#@/+,[:alnum:]:.-' '_' \
+                            | tr -s '_' \
+                            | sed 's/_$//' \
+                            )
+
+        case "X${test_case_result}" in
+            "XPASSED")
+                report_pass "${test_case_name}"
+                ;;
+            "XIGNORED")
+                report_skip "${test_case_name}"
+                ;;
+            "XFAILURE")
+                report_fail "${test_case_name}"
+                ;;
+            *)
+                report_unknown "${test_case_name}"
+                ;;
+        esac
+    done < "${f_kunit_stub_log}"
+}
+
+while getopts "u:s:h" o; do
+  case "$o" in
+    u) TESTS_ZIP_URL="${OPTARG}" ;;
+    s) SQUAD_UPLOAD_URL="${OPTARG}" ;;
+    h|*) usage ;;
+  esac
+done
+
+# export ANDROID_SERIAL
+initialize_adb
+
+if [ -z "${TESTS_ZIP_URL}" ]; then
+    echo "The TESTS_ZIP_URL must be specified."
+    exit 1
+fi
+
+# download and unzip tests.zip
+rm -f "${F_TESTS_ZIP}" && \
+    curl -fsSL "${TESTS_ZIP_URL}" -o "${F_TESTS_ZIP}"
+rm -fr "${DIR_TESTS}" && \
+    mkdir -p "${DIR_TESTS}" && \
+    unzip -o "${F_TESTS_ZIP}" -d "${DIR_TESTS}"
+
+# clone the tradefed prebuilts repository
+rm -fr "${DIR_TF_PREBUILTS}" && \
+    git clone --depth 1 "${TRADEFED_PREBUILTS_GIT_URL}" "${DIR_TF_PREBUILTS}"
```

Comment:
same here r.e. retry

Reply:
updated to be done in a while block, please help to check it again.